### PR TITLE
Improve Windows Compatibility(for lib/TH)

### PR DESCRIPTION
**Pull request by peterjc123**
Win64 support for lib/TH

### CODE REVIEW

File: torch/lib/TH/THAtomic.c
Comment (fmassa) [inline torch/lib/TH/THAtomic.c]: we can probably remove this comment

File: torch/lib/TH/THAtomic.c
Comment (fmassa) [inline torch/lib/TH/THAtomic.c]: this one as well

File: torch/lib/TH/THFile.c
Comment (fmassa) [inline torch/lib/TH/THFile.c]: `int64_t`

File: torch/lib/TH/generic/THBlas.h
Comment (fmassa) [inline torch/lib/TH/generic/THBlas.h]: nit: there are two spaces after `int64_t` everywhere in this file

File: torch/lib/TH/generic/THTensor.c
Comment (fmassa) [inline torch/lib/TH/generic/THTensor.c]: nit: is this the right format? I'm looking at https://stackoverflow.com/questions/9225567/how-to-print-a-int64-t-type-in-c and I'm not sure it's exactly what you are looking for. (This appears elsewhere in the codebase as well). Also, there are warning in the CI due to those changes, with warnings like
```
warning: format ‘%lld’ expects argument of type ‘long long int *’, but argument 3 has type ‘int64_t * {aka long int *}’ [-Wformat=]
```

File: torch/lib/TH/generic/THTensorRandom.c
Comment (fmassa) [inline torch/lib/TH/generic/THTensorRandom.c]: `_mask` is a `int64_t`, so no need for the casting I think

File: torch/lib/TH/THDiskFile.c
Comment (fmassa) [inline torch/lib/TH/THDiskFile.c]: There are warnings in the contribuild down in this file, maybe you miss some castings? Example of warning
```
/home/travis/build/pytorch/pytorch/torch/lib/TH/THDiskFile.c: In function ‘THDiskFile_new’:
/home/travis/build/pytorch/pytorch/torch/lib/TH/THDiskFile.c:618:5: warning: initialization from incompatible pointer type [-Wincompatible-pointer-types]
     THDiskFile_readChar,
     ^
```

**Review by fmassa — CHANGES_REQUESTED** [review]
Thanks for the PR!
I did a first pass on it, and had a look on the output of the contribuild. TH compiles, but the compiler outputs some warnings that might be good to fix before merging this PR.
Also, I didn't check the implementation inside `THAllocator.c`.

**Comment by peterjc123** [comment]
There 're no warnings now.

**Comment by fmassa** [comment]
Thanks for the changes!
TH seems now to compile without any warnings.
The build fails because of the dependency of THPP in TH, so we should probably only merge these PRs together.
@apaszke any better ideas on how we can run the tests with all the changes? The monolithic PR was too hard to review, the separate PRs are easier but running the full test suite is trickier.

**Comment by peterjc123** [comment]
@fmassa I tried to make CI builds with windows-full branch. Every time a change is made, it will be merged into this branch. Maybe it's useful.
For Linux/Mac    [Build Status](https://travis-ci.org/peterjc123/pytorch)
For Windows    [Build status](https://ci.appveyor.com/project/peterjc123/pytorch)

File: torch/lib/TH/generic/THTensorRandom.c
Comment (fmassa) [inline torch/lib/TH/generic/THTensorRandom.c]: We don't need the modulo here I suppose?

**Review by fmassa — APPROVED** [review]
Thanks a lot for your effort!
I think this is ready for merge. TH builds without warnings.
Let's wait until the other folders are reviewed so that we can merge them all.

**Comment by soumith** [comment]
Thanks a ton @peterjc123 for doing this big PR. This is really good stuff.

**Pull request merged**
